docs/features/role-based-access-control/high-level-design.md

```diff
@@ -0,0 +1,100 @@
+# Role base access control - High-level design
+
+This document aims at providing an overview of the architecture of Role-based Access Control (RBAC from here on) for
+the Marxan platform.
+
+## Roles
+
+Platform administrators should be identified by a boolean attribute in the users table (`is_admin`). This attribute can
+only be edited by other platform administrators, and it must not be editable by the user when editing their profile.
+
+The remaining roles supported by the platform (`Owner`, `Contributor` and `Viewer`) will be stored in the `roles` 
+database in the `marxan-api` database.
+
+Roles should have numerical values associated to them, to identify the hierarchical order between them. For instance:
+
+* `Owner` -> 1000
+* `Contributor` -> 500
+* `Viewer` -> 250
+
+Roles with fewer privileges should have lower numerical values than roles with more privileges. Using numerical values
+for roles allows for easily comparing if a given user has certain privileges (e.g. checking if a user has the role 
+`Contributor` can be easily done: `role >= Contributor (500)`).
+
+There is also an implicit relationship between platform actions and the supported roles. By default, it is assumed that:
+  * a user with role `Viewer` can read the information for the given resource (which usually includes calls to 
+`GET resource` and `GET resource/:id` endpoints).
+  * user with role `Contributor` can perform all the actions a `Viewer` can, as well as edition actions for the given
+resource (which usually includes calls to `PATCH/PUT resource/:id` endpoints).
+  * user with `Owner` role can perform all the actions a `Contributor` can, as well as deleting actions for the given
+resource (which usually includes calls to `DELETE resource/:id` endpoints).
+
+**However, it is up to each controller method to correctly check permissions for the action being performed.** This also applies
```

Comment:
Why are we restricting ourselves to controller level? I believe it may lead to multiple checks along the way.

Imagine, some role queries the Map Tiles, but depending on the role, they may be denied, or given different set of generated tiles depending on their scope (for example, hide some layers).

Another point to consider that it may end up in duplicating checks (and/or anyway moving them to some service/query) - imagine that some feature wants to trigger another action which may have different restrictions - we would either skip the check from controller, or have to call REST call, or duplicate it.

Also, this approach makes the testing difficult and binds the verification to very given HTTP transport layer.

Reply:
I didn't we would be **restricting** ourselves to controller level :)

I was using controller actions as examples (since it will very likely be the most common use case), but the main idea is that **the caller** (whatever that is, controller, service, function, etc) has the responsibility to correctly check permissions - in no way that is done automatically.

And the way to do it is by calling the service method from the permission module - `this.accessControlService.isOwner(resourceId, userId)` - a good example of this is how we are applying permissions to publish projects: https://github.com/Vizzuality/marxan-cloud/pull/610/files#diff-02daf12463952aecea26345f1a5364a7fcff93b7658cc1478c80e01bbde80c0aR40. In this case, the controller has the responsibility to check the permissions - this is what I meant with the sentence above.

Does this clarify your comment?

Reply:
`, but the main idea is that the caller`

it depends on who you meant to be the caller - controller rather shouldn't bother about it, it should be a service responsibility to ensure that other modules using some feature also include this check.

Reply:
Rephrased this section to make it clearer. However, the main point is of this sentence is: **permissions are not checked automatically**. They need to be explicitly verified in all places where we access resources. And the responsibility of doing that is on whoever implements the access to said resources. It's irrelevant where that access happens (controller, service, etc etc), but wherever it happens, permissions need to be checked explicitly by the caller.

Reply:
sounds good!